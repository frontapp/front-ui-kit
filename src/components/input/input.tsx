
import React, {FC, useLayoutEffect, useRef, ChangeEventHandler, FocusEventHandler, useState} from 'react';
import styled, {css} from 'styled-components';

import {Icon, IconName} from '../..';
import {greys, palette} from '../../helpers/colorHelpers';
import {fonts, fontSizes, fontWeights} from '../../helpers/fontHelpers';

/*
 * Constants.
 */

export enum InputTypesEnum {
  TEXT = 'text',
  NUMBER = 'number',
  EMAIL = 'email',
  PASSWORD = 'password'
}

/**
 * Props
 */

interface InputProps {
  /** The content of the input field */
  value: string | number;
  /** Type of input: number, text, email, url, password. Default to text. */
  type?: InputTypesEnum;
  /** The name of the input field */
  name?: string;
  /** Placeholder value */
  placeholder?: string;
  /** Whether the input is disabled */
  isDisabled?: boolean;
  /** The icon to render at the beginning of the input field */
  iconName?: IconName;
  /** Whether we should focus on the input when it is mounted */
  shouldFocus?: boolean;
  /** The handler for when the content of the input field changes */
  onChange: (value?: string | number) => void
  /** The handler for when the input field is unfocused */
  onBlur?: () => void
  /** The handler for when the input field is focused */
  onFocus?: () => void
}

/*
 * Style.
 */

interface StyledInputProps {
  $isDisabled: boolean;
  $isActive: boolean;
  $hasIcon: boolean;
}

const StyledInputDiv = styled.div`
  position: relative;
  width: 121px;
  height: 30px;
  display: flex;
  flex-flow: row;
`;

const StyledInput = styled.input<StyledInputProps>`
  width: inherit;
  height: inherit;
  background: ${greys.white};
  font-family: ${fonts.system};
  font-size: ${fontSizes.medium};
  font-weight: ${fontWeights.normal};
  line-height: 17px;
  text-overflow: ellipsis;
  border-radius: 8px;
  box-sizing: border-box;
  appearance: none;
  padding-left: ${p => p.$hasIcon ? `26px` : `5px`};
  border: none;
  outline: none;

  ${p => addInputStyles(p)};
`;

const StyledIconDiv = styled.div`
  position: absolute;
  top: 7px;
  left: 5px;
`;

function addInputStyles(props: StyledInputProps) {
  if (props.$isDisabled)
    return css`
      border: 2px solid ${greys.shade30};
      color: ${greys.shade70};
    `;

  if (props.$isActive)
    return css`
      border: 2px solid ${palette.blue.shade40};
      color: ${greys.shade90};
    `;

  return css`
    border: 2px solid ${greys.shade10};
    color: ${greys.shade90};
  `;
}

export const Input: FC<InputProps> = props => {
  const {value, type = InputTypesEnum.TEXT, name = "", isDisabled = false, iconName, onChange, onFocus, onBlur, shouldFocus = false} = props;
  const [isActive, setIsActive] = useState(shouldFocus);
  const inputRef = useRef<HTMLInputElement>(null);
  const onInputChange: ChangeEventHandler<HTMLInputElement> = event => {
    if (isDisabled || !onChange)
      return;
    const inputValue = event.currentTarget.value || '';
    onChange(inputValue);
  };

  const onInputFocus: FocusEventHandler<HTMLInputElement> = event => {
    if (isDisabled)
      return;
    const input = inputRef.current;
    if (!input)
      return;
    input.focus();
    setIsActive(true);
    if (onFocus)
      onFocus();
  };

  const onInputBlur: FocusEventHandler<HTMLInputElement> = event => {
    if (isDisabled)
      return;
    const input = inputRef.current;
    if (!input)
      return;
    input.blur();
    setIsActive(false);
    if (onBlur)
      onBlur();
  };

  // Set the attributes of the input on every render.
  useLayoutEffect(() => {
    const input = inputRef.current;
    if (!input)
      return;

    if (isDisabled)
      input.disabled = true;
    else
      input.disabled = false;
  }, [isDisabled]);

  return (
    <StyledInputDiv>
      <StyledIconDiv>{getInputIcon(iconName)}</StyledIconDiv>
      <StyledInput
        ref={inputRef}
        $isDisabled={isDisabled}
        $isActive={isActive}
        $hasIcon={Boolean(iconName)}
        autoFocus={shouldFocus}
        type={type}
        name={name}
        value={value}
        onChange={onInputChange}
        onFocus={onInputFocus}
        onBlur={onInputBlur}
      />
    </StyledInputDiv>
  );
};

/*
 * Helpers.
 */

function getInputIcon(iconName?: IconName) {
  if (iconName)
    return <Icon name={iconName}/>;
  return null;

}