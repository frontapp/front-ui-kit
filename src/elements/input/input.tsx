/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable react/function-component-definition */
/* eslint-disable @typescript-eslint/consistent-type-assertions */
import React, {ChangeEventHandler, FocusEventHandler} from 'react';
import styled, {css} from 'styled-components';

import {greys, palette} from '../../helpers/colorHelpers';
import {fonts, fontSizes, fontWeights} from '../../helpers/fontHelpers';
import {Icon, IconName} from '../icon/icon';

/*
 * Props
 */

interface InputProps<T> {
  /** The id of the input field */
  id?: string;
  /** The content of the input field */
  value?: T;
  /** Type of input: number, text, email, url, password. Default to text. */
  type?: 'text' | 'email' | 'password' | 'url' | 'number';
  /** The name of the input field */
  name?: string;
  /** Placeholder value */
  placeholder?: string;
  /** Whether the input is disabled */
  isDisabled?: boolean;
  /** The icon to render at the beginning of the input field */
  iconName?: IconName;
  /** Whether the input is erred */
  isErred?: boolean;
  /** Whether we should focus on the input when it is mounted */
  shouldFocus?: boolean;
  /** Max width of the input. Default is 100%. */
  maxWidth?: number;
  /** The handler for when the content of the input field changes */
  onChange?: (value: T) => void;
  /** The handler for when the input field is unfocused */
  onBlur?: FocusEventHandler;
  /** The handler for when the input field is focused */
  onFocus?: FocusEventHandler;
}

/*
 * Style.
 */

interface StyledInputDivProps {
  $maxWidth?: number;
}

const StyledInputDiv = styled.div<StyledInputDivProps>`
  position: relative;
  width: 100%;
  display: flex;
  flex-flow: row;

  ${(p) =>
    p.$maxWidth &&
    css`
      max-width: ${p.$maxWidth}px;
    `};
`;

interface StyledInputProps {
  $isDisabled: boolean;
  $hasIcon: boolean;
  $isErred: boolean;
}

const StyledInput = styled.input<StyledInputProps>`
  width: inherit;
  height: inherit;
  background: ${greys.shade20};
  font-family: ${fonts.system};
  font-size: ${fontSizes.medium};
  font-weight: ${fontWeights.normal};
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: ${(p) => (p.$hasIcon ? `25px` : `8px`)};
  outline: none;
  color: ${greys.shade90};

  &::placeholder {
    color: ${greys.shade60};
  }

  &:hover {
    background: ${greys.shade30};
  }

  &:focus {
    border: 2px solid ${palette.blue.shade40};
    background: ${greys.white};
  }

  ${(p) => addInputStyles(p)};
`;

const StyledIconDiv = styled.div`
  position: absolute;
  display: flex;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  justify-content: center;
`;

function addInputStyles(props: StyledInputProps) {
  if (props.$isDisabled)
    return css`
      border: 2px solid ${greys.shade30};
      background: ${greys.white};
      &:hover {
        background: ${greys.white};
      }
      opacity: 0.5;
    `;

  if (props.$isErred)
    return css`
      background: ${palette.red.shade10};
      border: 2px solid transparent;
      &:hover {
        background: ${palette.red.shade20};
      }
      &:focus {
        border: 2px solid ${palette.red.shade40};
        background: ${greys.white};
      }
    `;

  return css`
    border: 2px solid transparent;
  `;
}

/*
 * Component.
 */

export function Input<T = string>(props: InputProps<T>) {
  const {
    id,
    value,
    placeholder,
    type = 'text',
    name = '',
    isDisabled = false,
    iconName,
    isErred = false,
    maxWidth,
    onChange,
    onFocus,
    onBlur,
    shouldFocus = false
  } = props;
  const onInputChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    if (isDisabled || !onChange) return;
    const inputValue = event.currentTarget.value;
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    onChange(type === 'number' ? Number(inputValue) : inputValue);
  };

  const onInputFocus: FocusEventHandler<HTMLInputElement> = (event) => {
    if (isDisabled || !onFocus) return;
    onFocus(event);
  };

  const onInputBlur: FocusEventHandler<HTMLInputElement> = (event) => {
    if (isDisabled || !onBlur) return;
    onBlur(event);
  };

  return (
    <StyledInputDiv $maxWidth={maxWidth}>
      <StyledIconDiv>{getInputIcon(iconName)}</StyledIconDiv>
      <StyledInput
        id={id}
        $isDisabled={isDisabled}
        disabled={isDisabled}
        $isErred={isErred}
        $hasIcon={Boolean(iconName)}
        placeholder={placeholder}
        autoFocus={shouldFocus}
        type={type}
        name={name}
        value={value as unknown as string | number}
        onChange={onInputChange}
        onFocus={onInputFocus}
        onBlur={onInputBlur}
        aria-invalid={isErred}
      />
    </StyledInputDiv>
  );
}

/*
 * Helpers.
 */

function getInputIcon(iconName?: IconName) {
  if (iconName) return <Icon name={iconName} />;
  return null;
}
