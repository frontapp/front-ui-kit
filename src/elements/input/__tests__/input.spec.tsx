import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {Input} from '../input';

describe('<Checkbox /> Spec', () => {
  describe('Snapshots', () => {
    it('should render an empty and disabled input', () => {
      const view = render(<Input isDisabled id="id" placeholder="Placeholder" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render a filled and disabled input', () => {
      const view = render(<Input isDisabled id="id" value="Test" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an input with an error', () => {
      const view = render(<Input isErred id="id" value="Test" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an input with an icon', () => {
      const view = render(<Input iconName="CheckmarkBox" id="id" value="Test" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an input of type number', () => {
      const view = render(<Input iconName="CheckmarkBox" id="id" type="number" value={3} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an input of type email', () => {
      const view = render(<Input iconName="CheckmarkBox" id="id" type="email" value="hello@gmail.com" />);
      expect(view.container).toMatchSnapshot();
    });
  });

  it('should issue an onFocus when the input is clicked', async () => {
    const user = userEvent.setup();
    const onFocusMock = jest.fn();
    const view = render(<Input id="id" placeholder="Placeholder" onFocus={onFocusMock} />);
    await user.click(view.getByRole('textbox'));
    expect(onFocusMock).toBeCalled();
  });

  it('should issue an onBlur when some part of the screen is clicked after the input is focused', async () => {
    const user = userEvent.setup();
    const onBlurMock = jest.fn();
    const view = render(<Input id="id" placeholder="Placeholder" shouldFocus onBlur={onBlurMock} />);
    await user.click(view.container);
    expect(onBlurMock).toBeCalled();
  });

  it('should not issue an onFocus or an onBlur if the input is disabled', async () => {
    const user = userEvent.setup();
    const onFocusMock = jest.fn();
    const onBlurMock = jest.fn();
    const view = render(<Input id="id" isDisabled placeholder="Placeholder" onFocus={onFocusMock} onBlur={onBlurMock} />);
    await user.click(view.getByRole('textbox'));
    expect(onFocusMock).not.toBeCalled();
    await user.click(view.container);
    expect(onBlurMock).not.toBeCalled();
  });

  it('should issue an onChange when the user types into the input', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(<Input id="id" placeholder="Placeholder" onChange={onChangeMock} />);
    await user.type(view.getByRole('textbox'), "t");
    expect(onChangeMock).toBeCalled();
    expect(onChangeMock.mock.lastCall[0]).toBe("t");
  });

  it('should not issue an onChange when the user types into the input if the input is disabled', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(<Input id="id" isDisabled placeholder="Placeholder" onChange={onChangeMock} />);
    await user.type(view.getByRole('textbox'), "t");
    expect(onChangeMock).not.toBeCalled();
  });
});
