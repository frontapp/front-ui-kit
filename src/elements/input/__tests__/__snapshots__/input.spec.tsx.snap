// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Spec Snapshots should render a filled and disabled input 1`] = `
.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c2 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 8px;
  outline: none;
  color: #19191a;
  border: 2px solid #e8eaed;
  background: #ffffff;
  opacity: 0.5;
}

.c2::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c2::-moz-placeholder {
  color: #9b9c9e;
}

.c2:-ms-input-placeholder {
  color: #9b9c9e;
}

.c2::placeholder {
  color: #9b9c9e;
}

.c2:hover {
  background: #e8eaed;
}

.c2:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c2:hover {
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <input
      aria-invalid="false"
      class="c2"
      disabled=""
      id="id"
      name=""
      type="text"
      value="Test"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an empty and disabled input 1`] = `
.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c2 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 8px;
  outline: none;
  color: #19191a;
  border: 2px solid #e8eaed;
  background: #ffffff;
  opacity: 0.5;
}

.c2::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c2::-moz-placeholder {
  color: #9b9c9e;
}

.c2:-ms-input-placeholder {
  color: #9b9c9e;
}

.c2::placeholder {
  color: #9b9c9e;
}

.c2:hover {
  background: #e8eaed;
}

.c2:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c2:hover {
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <input
      aria-invalid="false"
      class="c2"
      disabled=""
      id="id"
      name=""
      placeholder="Placeholder"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an input of type email 1`] = `
.c2 {
  width: 16px;
  height: 16px;
}

.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c3 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 25px;
  outline: none;
  color: #19191a;
  border: 2px solid transparent;
}

.c3::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c3::-moz-placeholder {
  color: #9b9c9e;
}

.c3:-ms-input-placeholder {
  color: #9b9c9e;
}

.c3::placeholder {
  color: #9b9c9e;
}

.c3:hover {
  background: #e8eaed;
}

.c3:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          color="#bbbdbf"
          height="16"
          name="CheckmarkBox"
          preserveaspectratio="none"
          width="16"
        />
      </div>
    </div>
    <input
      aria-invalid="false"
      class="c3"
      id="id"
      name=""
      type="email"
      value="hello@gmail.com"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an input of type number 1`] = `
.c2 {
  width: 16px;
  height: 16px;
}

.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c3 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 25px;
  outline: none;
  color: #19191a;
  border: 2px solid transparent;
}

.c3::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c3::-moz-placeholder {
  color: #9b9c9e;
}

.c3:-ms-input-placeholder {
  color: #9b9c9e;
}

.c3::placeholder {
  color: #9b9c9e;
}

.c3:hover {
  background: #e8eaed;
}

.c3:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          color="#bbbdbf"
          height="16"
          name="CheckmarkBox"
          preserveaspectratio="none"
          width="16"
        />
      </div>
    </div>
    <input
      aria-invalid="false"
      class="c3"
      id="id"
      name=""
      type="number"
      value="3"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an input with an error 1`] = `
.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c2 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 8px;
  outline: none;
  color: #19191a;
  background: #fff1f1;
  border: 2px solid transparent;
}

.c2::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c2::-moz-placeholder {
  color: #9b9c9e;
}

.c2:-ms-input-placeholder {
  color: #9b9c9e;
}

.c2::placeholder {
  color: #9b9c9e;
}

.c2:hover {
  background: #e8eaed;
}

.c2:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c2:hover {
  background: #ffe1de;
}

.c2:focus {
  border: 2px solid #e9483a;
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <input
      aria-invalid="true"
      class="c2"
      id="id"
      name=""
      type="text"
      value="Test"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an input with an icon 1`] = `
.c2 {
  width: 16px;
  height: 16px;
}

.c0 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
}

.c3 {
  width: inherit;
  height: inherit;
  background: #f3f4f6;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  border-radius: 6px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 4.5px 8px 4.5px 8px;
  padding-left: 25px;
  outline: none;
  color: #19191a;
  border: 2px solid transparent;
}

.c3::-webkit-input-placeholder {
  color: #9b9c9e;
}

.c3::-moz-placeholder {
  color: #9b9c9e;
}

.c3:-ms-input-placeholder {
  color: #9b9c9e;
}

.c3::placeholder {
  color: #9b9c9e;
}

.c3:hover {
  background: #e8eaed;
}

.c3:focus {
  border: 2px solid #367fee;
  background: #ffffff;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  top: 0;
  bottom: 0;
  left: 5px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          color="#bbbdbf"
          height="16"
          name="CheckmarkBox"
          preserveaspectratio="none"
          width="16"
        />
      </div>
    </div>
    <input
      aria-invalid="false"
      class="c3"
      id="id"
      name=""
      type="text"
      value="Test"
    />
  </div>
</div>
`;
